src/btree/bt_sync.c

```diff
@@ -223,9 +223,12 @@ __sync_ref_obsolete_check(WT_SESSION_IMPL *session, WT_REF *ref)
     if (busy)
         return (0);
 
+    /* Skip the modified pages as their reconciliation results are not valid any more. */
     WT_ASSERT(session, ref->page != NULL);
-    mod = ref->page->modify;
+    if (__wt_page_is_modified(ref->page))
+        goto err;
 
+    mod = ref->page->modify;
```

Comment:
Can we do this check early on in the function (it's just another `Ignore` clause), and avoid both locking the WT_REF and acquiring the hazard pointer on modified pages?

Reply:
I thought of doing it as a first check, but I feel there may be some race conditions that can lead to a situation where `page` can be set to NULL after we checked and it can lead to an abort? As per my understanding, we need to lock the REF or take a hazard pointer before deciding whether the page is modified or not.

Reply:
I think you're correct -- absent the hazard pointer, the page could be evicted and `WT_REF.page` goes to NULL.

I withdraw the comment, although you might want to note the requirement in the change for clarity. (No strong feelings, whatever you think best.)

Reply:
Thanks. I will add a comment for better clarity.

Reply:
Fixed.